### ISSUE TITLE
Telnet support

### ISSUE THREAD

**Issue by sivabudh** [ktbyers/netmiko#159]
Hello,

Is there a way to use Telnet instead of SSH?

PS: Awesome library. :+1:

**Comment by ktbyers** [comment]
Thanks @sivabudh 

Yes, I don't think I am going to add telnet support. Too much work and too marginal a need.

I will keep this issue open for a while so if there are other people that are interested in this--they can voice their opinion.

**Comment by calaesha** [comment]
Telnet would allow us to build an out-of-the-box switch or router with our Python apps. Having to console in, create domain-name, hostname, crypto keys, etc delays the automation process so much that it's almost moot.

May be a niche case, but I'd enjoy seeing it.

**Comment by ktbyers** [comment]
@calaesha Which vendor/platform are you working on? Can you explain your workflow a bit more (or the workflow you want)?

If it is Cisco ISR with one-time password doesn't SSH work without config (definitely not sure about this)?

If it is not ISR with one-time password, wouldn't you have to do some config to get telnet to work (at least add a VTY password or configure AAA with a username)?

Some other solutions to the above problem (of automatically bootstrapping a working SSH):
1. pyserial (https://pynet.twb-tech.com/blog/automation/pyserial.html)
2. A terminal server that allows SSH and then script to it
3. Just write a small telnet script to do the SSH bootstrap

I would probably help with item #3 if you are interested in doing that (assuming Cisco devices).

**Comment by calaesha** [comment]
Hello @ktbyers, the dream would be to have a non-technical individual run a python script against a brand new Cisco switch or router. 
- No config within the NVRAM will initiate Cisco's AutoInstall feature: a DHCP request is broadcast'd out all Ethernet interfaces (RARP across serial)
- assign the address with an 'interface x \ address dhcp'
- junior team member would initiate the python configuration builder script with netmiko over telnet

Across all hardware platforms, the device will reject all SSH sessions until:
- hostname is configured
- domain-name is configured
- crypto keys generated
- ssh version announced

some other tangibles that come into play are:
- line vty ports
- aaa authentication
- username / password generated

In short, there is a significant effort to ramping up pyserial on a non-technical user's workstation (drivers, com port assignments, etc). 

Are you suggesting in your third suggestion to try a combination of TELNET+Pexpect and NetMiko in the same app?

**Comment by ktbyers** [comment]
@calaesha 

Goal as I see it is zero manual config (obviously closer to this is better i.e. manually configuring one line is better than four etc.).

What I was suggesting is maybe I can add a small bootstrap telnet driver to Netmiko (for Cisco devices). The bootstrap telnet driver would be enough to let you configure SSH. But would not be intended for full telnet support (i.e. if you don't have or not using SSH on these network devices, then you would be out of luck).

You could have one script, but it would switch from telnet to SSH after SSH was configured (i.e. two separate connections).

If you want to work on a proof of concept, I am probably open to it.

**Comment by calaesha** [comment]
Sounds great, let me know what I can do to help out

**Comment by nu110** [comment]
I haven't tried it yet, but I wonder if ConsolServer (https://pypi.python.org/pypi/ConsoleServer) might help, looks like an implementation of Ser2Net which uses ssh rather than telnet.

**Comment by sivabudh** [comment]
@ktbyers, what if I were to submit a pull request that will allow netmiko's users to specify the network protocols like this:

```
# By default it's SSH
>>> cisco_881 = {
     'device_type': 'cisco_ios',
     'ip':   '10.10.10.10',
     'username': 'test',
     'password': 'password',
}

# But, they can also use telnet as well
>>> huawei_dev = {
     'device_type': 'huawei',
     'protocol':   'telnet',                 # optional, defaults to 'ssh'
     'ip':   '10.10.10.11',
     'username': 'test',
     'password': 'password',
}
```

I think there are benefits to support both telnet and ssh because existing networks (with thousands of network devices) might not have been configured with SSH. Thoughts?

**Comment by ktbyers** [comment]
@sivabudh 

Yes, I would prefer the following standard:

cisco_ios_ssh

I.E. vendor_platform_protocol. I already added code that will append "_ssh" to all device types to add protocol to the end of the device type. If protocol is not specified i.e. 'cisco_ios', then SSH will be assumed.

Yes, we would want to use telnetlib to implement.

If we want to do this, then we should mock out support. It would be nice to leverage as much existing code as we could (i.e. it would be nice to avoid having to duplicate a large amount of code where we do all of the same things--one with the SSH driver and one with the telnet driver).

We could possibly create a branch where we proof of concept this.

**Comment by ktbyers** [comment]
FYI, I started working on this here:

https://github.com/ktbyers/netmiko/tree/telnet_poc

I would like to add support for the following (need to schedule a week to work on this):

_ssh
_telnet
_serial

So you could call

cisco_ios_ssh
cisco_ios_telnet
cisco_ios_serial

Where cisco_ios_ could be any netmiko vendor_platform

If anyone wants to work on this, let me know.


Kirk

**Comment by ktbyers** [comment]
@calvinchristopher Let me know if I left you hanging here as I never did the proof-of-concept code.

**Comment by calvinchristopher** [comment]
Hello Kirk, I apologize for missing your message. I would love to give this a spin, please let me know what you need from my side

**Comment by ktbyers** [comment]
@calvinchristopher I will probably try to integrate telnet support into Netmiko. Not sure when as fairly big project. I will post on this thread when for additional people to test.

**Comment by sivabudh** [comment]
@ktbyers, my project requires telnet support. I'm actually looking into adding telnet into Netmiko right now.

**Comment by sivabudh** [comment]
@ktbyers, this is a very preliminary work.

https://github.com/sivabudh/python-telnet/blob/master/main.py

I'm testing with a Huawei S2300 right now.

**Comment by sivabudh** [comment]
@ktbyers, I'm done with my telnet script that can communicate with Huawei now.

The code is still in the same place.
https://github.com/sivabudh/python-telnet/blob/master/main.py

I studied your code, and made a more or less a stripped-down version of what you had.

Basically, my script can send commands to Huawei S2300 via Telnet with some minor features:
1. It handles detecting invalid login attempt
2. It figures out the network prompt at run-time, and does support the case when the prompt changes due to mode change (such as being changed to configuration mode)
3. It does not use delay in order to determine whether the network device has completely returned output. Instead, it looks for the expected prompt. This is acceptable for me because the script will be executed in the background by Python's `rq` worker where I can specify a timeout and terminates the process should the script hang: http://python-rq.org/docs/ (search for `timeout` on the page).

Next, I need to refactor this code a bit as it's still a bit messy now. I will also need to create a `Juniper` module as I intend to use this code with Juniper devices as well.

My priority is to get my project done first, and I will see how/if you would be open to my pull request(s). (If I could figure out how to integrate it well into `netmiko`, of course.)

Cheers!

**Comment by lamehost** [comment]
+1 for telnet support

**Comment by jwbensley** [comment]
+1 for telnet support from me to, we could add this to the NAPALM Hackathon list or is it too non-NAPALM specific?

**Comment by ktbyers** [comment]
@jwbensley Yes, I think this is probably too Netmiko specific for the NAPALM hackathon.

**Comment by sivabudh** [comment]
Updates: I have been running my own Telnet port which is able to connect to Juniper and Huawei fine. Still haven't got the time to figure out how to merge this into `netmiko` though. :(

**Comment by ktbyers** [comment]
This is being worked on here:  https://github.com/ktbyers/netmiko/issues/245

**Comment by ktbyers** [comment]
Okay, telnet driver has been added for Cisco IOS into Netmiko dev_1_0 branch. Feel free to test it. If you want support for other platforms, you will need to commit to testing it (and I can possibly/probably implement it as long as there is a current Netmiko SSH driver).

**Comment by jeffsilverm** [comment]
I would like telnet support for Juniper.  I have 1 lame use case and 1 good use case.  I am writing some test automation, and one of the requirements is to login to the serial console.  I have a solution now which uses telnetlib, and it works but it is not reliable.  The other requirement is that I am telnet directly.  The use case that the sshd has died or there is a firewall in the network that blocks port 22.

**Comment by ktbyers** [comment]
@jeffsilverm Yes, feel to implement it (i.e. submit a PR on it). You can see what I did here for Cisco.

https://github.com/ktbyers/netmiko/blob/develop/netmiko/cisco_base_connection.py#L52

I don't know if I can bring myself to write more telnet code  :-)

**Comment by magic9669** [comment]
Is there telnet support for XR devices?  It worked with IOS using the cisco_ios_telnet device type but if i try to do the cisco_xr_telnet device type, it doesn't seem to work.

**Comment by magic9669** [comment]
While i'm at it, is there support for NX-OS, Juniper and EOS devices as well for telnet?  I work for Comcast, but in their lab environment and deal with hundreds of devices from multiple vendors.  Within their lab environment, unless its public facing, everything is telnet, that's why I ask.  Thanks all.

**Comment by ktbyers** [comment]
I only have telnet support for Cisco IOS (there might be one other platform). I am open to PRs on other platforms, but I am probably not going to write the telnet support for the other platforms.

I can help point in the right direction and get the structure setup in a reasonable way.

**Comment by magic9669** [comment]
Sure.  I wouldn't mind learning something new.  I'm VERY new to Python and netmiko but any assistance you could provide, I would love trying to get it to work myself.  Thanks Keith

**Comment by ranjeetjangra** [comment]
Hi ,
I need Telnet support for Alcatel     
"alcatel_aos": AlcatelAosSSH,
"alcatel_sros": AlcatelSrosSSH,

**Comment by ktbyers** [comment]
@ranjeetjangra Feel free to submit a PR on it. You can look at other telnet drivers for the patterns to implement.

**Comment by boxie2-cisco** [comment]
Does netmiko can support telnet for cisco_xr platform now?

**Comment by ktbyers** [comment]
Yes:

https://github.com/ktbyers/netmiko/blob/develop/netmiko/ssh_dispatcher.py#L241

**Comment by boxie2-cisco** [comment]
Many thanks kirk. Good solution, perfect.

Xie Bo
ENGINEER.NETWORK CONSULTING
Cisco Services
boxie2@cisco.com<mailto:boxie2@cisco.com>
Phone: +86 10 8515 5757
Mobile: +86 13911801476

Cisco Systems (China) Information Technology Services Limited
7-12F Building C, Yintai Office Tower, Beijing Yintai Centre
No. 2 Jianguomenwai Ave, Chaoyang District
100022
Beijing
China

**Comment by boxie2-cisco** [comment]
Hello, Kirk:
I tested the netmiko  and using telnet  to login cisco XR device, but failed. Can you help me to check it ? thanks
Bellowing is log:
>>> from netmiko import Netmiko
>>> net_connect = Netmiko(host="210.52.128.12", username="T-SK-cisconos", password="8888#", device_type="cisco_xr_telnet")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.8/site-packages/netmiko/ssh_dispatcher.py", line 302, in ConnectHandler
    return ConnectionClass(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/netmiko/base_connection.py", line 344, in __init__
    self._open()
  File "/usr/local/lib/python3.8/site-packages/netmiko/base_connection.py", line 349, in _open
    self.establish_connection()
  File "/usr/local/lib/python3.8/site-packages/netmiko/cisco/cisco_xr.py", line 9, in establish_connection
    super().establish_connection(width=511, height=511)
  File "/usr/local/lib/python3.8/site-packages/netmiko/base_connection.py", line 908, in establish_connection
    self.telnet_login()
  File "/usr/local/lib/python3.8/site-packages/netmiko/cisco_base_connection.py", line 169, in telnet_login
    raise NetmikoAuthenticationException(msg)
netmiko.ssh_exception.NetmikoAuthenticationException: Login failed: 210.52.128.12
>>>
...
>>>
[root@fb1 xxx]#  telnet 210.52.128.12
Trying 210.52.128.12...
Connected to 210.52.128.12.
Escape character is '^]'.

username: T-SK-cisconos
password:
Last switch-over Thu Jan 28 18:23:56 2016: 4 years, 32 weeks, 2 days, 2 hours, 18 minutes ago



Xie Bo
ENGINEER.NETWORK CONSULTING
Cisco Services
boxie2@cisco.com<mailto:boxie2@cisco.com>
Phone: +86 10 8515 5757
Mobile: +86 13911801476

Cisco Systems (China) Information Technology Services Limited
7-12F Building C, Yintai Office Tower, Beijing Yintai Centre
No. 2 Jianguomenwai Ave, Chaoyang District
100022
Beijing
China


From: boxie2 boxie <boxie2@cisco.com>
Date: Thursday, September 3, 2020 at 2:13 PM
To: ktbyers/netmiko <reply@reply.github.com>, ktbyers/netmiko <netmiko@noreply.github.com>
Cc: Comment <comment@noreply.github.com>
Subject: Re: [ktbyers/netmiko] Telnet support (#159)

Many thanks kirk. Good solution, perfect.

Xie Bo
ENGINEER.NETWORK CONSULTING
Cisco Services
boxie2@cisco.com<mailto:boxie2@cisco.com>
Phone: +86 10 8515 5757
Mobile: +86 13911801476

Cisco Systems (China) Information Technology Services Limited
7-12F Building C, Yintai Office Tower, Beijing Yintai Centre
No. 2 Jianguomenwai Ave, Chaoyang District
100022
Beijing
China
